File: CodeSamples/kahn.jl
function kahn(graph, n)
    inverse = Vector{Vector{Int}}(undef, n)
    fill!(inverse, [])
    for u in 1:n
        for v in graph[u]
            push!(inverse[v], u)
        end
    end
    stack = Int[]
    for v in 1:n
        if isempty(inverse[v])
            push!(stack, v)
        end
    end
    order = Int[]
    while !isempty(stack)
        u = pop!(stack)
        push!(order, u)
        for v in graph[u]
            deleteat!(graph[u], findfirst(isequal(v), graph[u]))
            deleteat!(inverse[v], findfirst(isequal(u), inverse[v]))
            if isempty(inverse[v])
                push!(stack, v)
            end
        end
    end
    if length(order) != n
        println("Cyclic graph")
    else
        println(join(order, " "))
    end
end

input_file = open("input.txt", "r")
output_file = open("output.txt", "w")
T = parse(Int, readline(input_file))
for t in 1:T
    n = parse(Int, readline(input_file))
    graph = Vector{Vector{Int}}(undef, n)
    fill!(graph, Int[])
    while true
        line = readline(input_file)
        if isempty(line)
            break
        end
        u, v = parse.(Int, split(line))
        push!(graph[u], v)
    end
    kahn(graph, n)
end
close(input_file)
close(output_file)

File: CodeSamples/kahn.pl
use Data::Dumper;

sub kahn {
  my ($graph, $n) = @_;
  my $inverse = {};
  for my $u (@$graph) {
    for my $v (@$graph->{$u}) {
      $inverse->{$v} //= [];
      push $inverse->{$v}, $u;
    }
  }

  my @stack = grep { !exists $inverse->{$_} }, (0 .. $n - 1);
  my @order = ();
  while (@stack) {
    my $u = pop @stack;
    push @order, $u;

    for my $v (@$graph->{$u}) {
      splice @$graph->{$u}, grep { $_ == $v }, 1;
      splice @$inverse->{$v}, grep { $_ == $u }, 1;
      if (!exists $inverse->{$v}) {
        push @stack, $v;
      }
    }
  }

  if (@order != $n) {
    return "Cyclic graph";
  }
  return @order;
}

open INPUT, "<input.txt";
open OUTPUT, ">output.txt";

my $n = <INPUT>;
my $graph = {};
while
